tiledb/api/src/array/dimension/mod.rs

```diff
@@ -164,101 +163,219 @@ impl<'c1, 'c2> PartialEq<Dimension<'c2>> for Dimension<'c1> {
+                    DimensionConstraints::$V([value.0[0], value.0[1]], value.1)
+                }
+            }
+        )+
     }
+}
 
-    /// N.B., This API will eventually be removed when we fixup the current
-    /// implementation of `new` to take a `DimensionConstraint` argument instead
-    /// of the serde values for range and extent.
-    pub fn new_string(
-        context: &'ctx Context,
-        name: &str,
+impl From<()> for DimensionConstraints {
+    fn from(_value: ()) -> DimensionConstraints {
+        DimensionConstraints::StringAscii
+    }
+}
+
+dimension_constraints_impl!(Int8: i8, Int16: i16, Int32: i32, Int64: i64);
+dimension_constraints_impl!(UInt8: u8, UInt16: u16, UInt32: u32, UInt64: u64);
+dimension_constraints_impl!(Float32: f32, Float64: f64);
+
+impl DimensionConstraints {
+    pub fn verify_type_compatible(
+        &self,
         datatype: Datatype,
-    ) -> TileDBResult<Self> {
-        let c_datatype = datatype.capi_enum();
-        let c_name = cstring!(name);
-        let mut c_dimension: *mut ffi::tiledb_dimension_t =
-            std::ptr::null_mut();
+    ) -> TileDBResult<()> {
+        dimension_constraints_go!(
+            self,
+            DT,
+            _range,
+            _extent,
+            {
+                if !datatype.is_compatible_type::<DT>() {
+                    return Err(Error::Datatype(
+                        DatatypeErrorKind::TypeMismatch {
+                            user_type: std::any::type_name::<DT>(),
+                            tiledb_type: datatype,
+                        },
+                    ));
+                }
+            },
+            {
+                if !matches!(datatype, Datatype::StringAscii) {
+                    return Err(Error::Datatype(
+                        DatatypeErrorKind::InvalidDatatype {
+                            context: Some(
+                                "DimensionConstraints::StringAscii".to_owned(),
+                            ),
+                            found: datatype,
+                            expected: Datatype::StringAscii,
+                        },
+                    ));
+                }
+            }
+        );
 
-        context.capi_call(|ctx| unsafe {
-            ffi::tiledb_dimension_alloc(
-                ctx,
-                c_name.as_ptr(),
-                c_datatype,
-                std::ptr::null(),
-                std::ptr::null(),
-                &mut c_dimension,
-            )
-        })?;
-        Ok(Builder {
-            dim: Dimension {
-                context,
-                raw: RawDimension::Owned(c_dimension),
+        Ok(())
+    }
+
+    pub fn domain_ptr(&self) -> *const std::ffi::c_void {
+        dimension_constraints_go!(
+            self,
+            DT,
+            range,
+            _extent,
+            range.as_ptr() as *const DT as *const std::ffi::c_void,
+            std::ptr::null()
```

Comment:
Can these return `Option<*const std::ffi::c_void>`?  Feels better (to me) than checking for null pointer.

`std::ptr::NonNull` looks interesting too but it assumes `*mut` which isn't appropriate here.

Also these should probably be `pub(crate)` instead of `pub`, IMO we shouldn't leak any FFI types.

Reply:
Good point that this is not checked but passed straight to the FFI call.  Option would just be a headache then.